python/thunder/imgprocessing/transformation.py

```diff
@@ -8,6 +9,168 @@ def apply(self, im):
+            ordered changes in parameters to be applied. The ordering here will match
+            the ordering of the parameters returned by jacobian.
+        """
+        raise NotImplementedError
+
+
+class AffineTransformation(DifferentiableTransformation):
+    """ Affine transformations are differentiable and can be represented as a matrix."""
+    def matrix(self):
+        raise NotImplementedError
+
+    def apply(self, vol, grid=None, order=1):
+        if grid is None:
+            center = getattr(self, 'center', None)
+            grid = GridTransformer(vol.shape, center=center)
+        A = self.matrix()
+        return grid.transform_vol(vol, A, order=order)
+
+
+class TranslationTransformation(AffineTransformation):
+    def __init__(self, shift):
+        """Translation in 3d.
+
+        Parameters
+        ----------
+            shift: 3d vector of translations (x, y, z)
+
+        """
+        self.delta = shift
+
+    def jacobian(self, imageGradients, imageGrid):
+        return imageGradients
+
+    def matrix(self):
+        return transformationMatrix(self.delta)
+
+    def updateParams(self, deltaParams):
+        self.delta += deltaParams
+
+class EuclideanTransformation(AffineTransformation):
+    def __init__(self, shift, rotation=None, zTranslation=False, zRotation=False, center=None):
+        """Translation and rotation in 3d.
+
+        Parameters
+        ----------
+        shift : list or array
+            spatial shifts for each dimension
+        rotation : float, list, or array
+            rotation in x-y if scalar. rotation in x-y plane, x-z plane, y-z plane if list
+            or array and dataset is in 3d.
+        zTranslation : bool, optional, default = False
+            whether to allow translation in z
+        zRotation : bool, optional, default = False
+            whether to allow rotation in z
+        center : list or array, optional, default = None
+            Set center coordinates that define the point around which the volume is rotated.
+            Coordinates should be in terms of zero-indexed pixels. For example if the volume was 5x5x3,
+            the default center of rotation would be at the center of the volume: (2, 2, 1).
+        """
+        self.shift = np.atleast_1d(shift)
+        self.ndim = len(self.shift)
+        if rotation is None:
+            if self.ndim == 2 or not zRotation:
+                rotation = 0.0
+            else:
+                rotation = np.zeros(3)
+        self.rotation = np.atleast_1d(rotation)
+        self.zRotation = zRotation
+        self.zTranslation = zTranslation
+        self.center = center
+
+    def updateParams(self, deltaParams):
+        if self.ndim == 2:
+            self.shift += deltaParams[:2]
+            self.rotation += deltaParams[2]
+        else:
+            self.shift += deltaParams[:3]
+            self.rotation += deltaParams[3:]
+
+    def jacobian(self, imageGradients, imageGrid):
```

Comment:
Confusing to use `imagaGradients` as the var name because that's the name of the function defined in a related module for related operations.